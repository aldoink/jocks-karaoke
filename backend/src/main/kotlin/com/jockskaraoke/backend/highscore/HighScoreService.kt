package com.jockskaraoke.backend.highscore

import org.springframework.stereotype.Service
import javax.transaction.Transactional

@Service
@Transactional
class HighScoreService(
    private val highScoreRepository: HighScoreRepository
) {
    fun findHighscores(songId: Int): List<HighScore> {
        return highScoreRepository.findAllBySongId(songId)
    }

    fun saveHighScore(name: String, score: Int, songId: Int) {
        highScoreRepository.save(HighScore(0, name, score, songId))
    }

    fun deleteHighScore(highScoreId: Int) {
        highScoreRepository.deleteById(highScoreId)
    }
}