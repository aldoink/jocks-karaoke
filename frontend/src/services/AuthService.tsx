import axios from "axios";

export const BACKEND_URL = process.env.REACT_APP_BACKEND_URL;

export class AuthService {
    async login (email: string, password: string): Promise<void> {
        const response = await axios.post(`${BACKEND_URL}/auth/login`, {email, password});
        localStorage.setItem("token", response?.data?.token);
    }

    getToken() {
        return localStorage.getItem("token")
    }
}