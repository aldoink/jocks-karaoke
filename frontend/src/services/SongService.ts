import axios from "axios";

export class SongService {
    private baseUrl = process.env.REACT_APP_BACKEND_URL;

    public async search(searchTerm: string) {
        let requestUrl = `${this.baseUrl}/songs`;
        if (searchTerm) {
            requestUrl += `?searchTerm=${searchTerm}`;
        }
        const response = await axios.get(requestUrl);
        return response?.data;
    }
}