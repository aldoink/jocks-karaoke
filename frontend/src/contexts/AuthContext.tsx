import {createContext} from "react";
import {AuthService} from "../services/AuthService";

export const AuthContext = createContext({authService: {} as AuthService});