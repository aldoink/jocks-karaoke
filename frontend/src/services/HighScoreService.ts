import axios, {AxiosRequestConfig} from "axios";
import {AuthService, BACKEND_URL} from "./AuthService";

export interface HighScore {
    [index: string]: string | number | undefined;

    name?: string;
    score?: string;
    songId: number;
}

export class HighScoreService {
    constructor(private authService: AuthService) {
    }

    public async findAll(songId: number): Promise<HighScore[]> {
        const response = await axios.get(`${BACKEND_URL}/highscores/${songId}`, this.getAxiosConfig());
        return response?.data;
    }

    public async save(entry: HighScore) {
        return await axios.put(
            `${BACKEND_URL}/highscores`,
            entry,
            this.getAxiosConfig()
        )
    }

    private getAxiosConfig(): AxiosRequestConfig {
        return {headers: {Authorization: "Bearer " + this.authService.getToken()}};
    }
}