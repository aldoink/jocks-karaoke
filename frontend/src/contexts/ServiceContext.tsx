import {createContext} from "react";
import {HighScoreService} from "../services/HighScoreService";
import {SongService} from "../services/SongService";

export interface IServiceContext {
    readonly highScoreService: HighScoreService
    readonly songService: SongService
}

export const ServiceContext = createContext<IServiceContext>(
    {
        highScoreService: {} as HighScoreService,
        songService: {} as SongService
    }
);

