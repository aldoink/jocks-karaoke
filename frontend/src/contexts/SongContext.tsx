import { createContext, Dispatch, SetStateAction } from "react";
import { SongService } from "../services/SongService";
import { Song } from "../models/Song";

export const songService = new SongService();
export interface ISongContext {
  readonly songService: SongService;
  songs: Song[];
  setSongs: Dispatch<SetStateAction<Song[]>>;
}

export const SongContext = createContext<ISongContext>({
  songService,
  songs: [],
  setSongs: () => {},
});
