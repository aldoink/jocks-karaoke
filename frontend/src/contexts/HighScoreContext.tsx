import { createContext } from "react";
import { HighScore, HighScoreService } from "../services/HighScoreService";
import { authService } from "./AuthContext";
import { Song } from "../models/Song";

export const highScoreService = new HighScoreService(authService);

export interface IHighScoreContext {
  highScoreService: HighScoreService;
  highScores: HighScore[];
  refreshHighScores: (song: Song) => void;
  hasError: boolean;
}

export const HighScoreContext = createContext<IHighScoreContext>({
  highScoreService,
  highScores: [],
  refreshHighScores: () => {},
  hasError: false,
});
