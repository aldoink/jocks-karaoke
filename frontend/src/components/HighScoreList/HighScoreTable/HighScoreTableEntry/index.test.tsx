import {HighScoreTableEntry} from "./index";
import {HighScore, HighScoreService} from "../../../../services/HighScoreService";
import {render} from "@testing-library/react";
import {screen} from "@testing-library/dom";
import {IServiceContext, ServiceContext} from "../../../../contexts/ServiceContext";
import userEvent from "@testing-library/user-event";

describe('HighScoreTableEntry', () => {
    const defaultHighScore: HighScore = {name: 'Test', score: 99, songId: 123};
    const mockedHighScoreService = {} as HighScoreService;

    const renderHighScoreTableEntry = (entry: HighScore = defaultHighScore, editMode: boolean = false) => render(
        <ServiceContext.Provider value={{highScoreService: mockedHighScoreService} as IServiceContext}>
            <table>
                <tbody>
                <HighScoreTableEntry entry={entry} editMode={editMode}/>
                </tbody>
            </table>
        </ServiceContext.Provider>
    )

    function saveButton() {
        return screen.getByText('Save');
    }

    beforeEach(() => {
        mockedHighScoreService.save = jest.fn();
    });

    it('renders correctly when editMode = false', () => {
        expect(renderHighScoreTableEntry().asFragment())
            .toMatchSnapshot();
        expect(screen.queryByTestId('name-input')).not.toBeInTheDocument();
        expect(screen.queryByTestId('score-input')).not.toBeInTheDocument();
        expect(screen.getByText(defaultHighScore.name)).toBeInTheDocument();
        expect(screen.getByText(defaultHighScore.score)).toBeInTheDocument();
    });

    function nameInputField() {
        return screen.getByTestId('name-input');
    }

    function scoreInputField() {
        return screen.getByTestId('score-input');
    }

    it('renders correctly when editMode = true', () => {
        //given
        expect(renderHighScoreTableEntry(defaultHighScore, true).asFragment())
            .toMatchSnapshot();

        //then
        expect(nameInputField()).toBeInTheDocument();
        expect(scoreInputField()).toBeInTheDocument();
        expect(screen.queryByText(defaultHighScore.name)).not.toBeInTheDocument();
        expect(screen.queryByText(defaultHighScore.score)).not.toBeInTheDocument();
        expect(saveButton()).toBeInTheDocument();
    });

    it('saves the high score when save button is clicked', () => {
        //given
        renderHighScoreTableEntry(defaultHighScore, true);

        //when
        userEvent.click(saveButton());

        //then
        expect(mockedHighScoreService.save).toHaveBeenCalledWith(defaultHighScore);
    });

    it('saves changes to the current entry when save button is clicked', async () => {
        //given
        renderHighScoreTableEntry(defaultHighScore, true);
        const name = 'Someone new';
        const score = '88';
        const expectedHighScore: HighScore = {name, score: 88, songId: defaultHighScore.songId};

        //when
        await userEvent.clear(nameInputField());
        userEvent.type(nameInputField(), name);
        await userEvent.clear(scoreInputField());
        userEvent.type(scoreInputField(), score);

        userEvent.click(saveButton())

        //then
        expect(mockedHighScoreService.save).toHaveBeenCalledWith(expectedHighScore);
    });
});