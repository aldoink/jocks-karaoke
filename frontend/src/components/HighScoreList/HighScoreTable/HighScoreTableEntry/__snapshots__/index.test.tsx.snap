// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HighScoreTableEntry renders correctly when editMode = false 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="table-cell"
        >
          Test
        </td>
        <td
          class="table-cell"
        >
          99
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`HighScoreTableEntry renders correctly when editMode = true 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="table-cell"
        >
          <input
            data-testid="name-input"
            type="text"
            value="Test"
          />
        </td>
        <td
          class="table-cell"
        >
          <input
            data-testid="score-input"
            type="number"
            value="99"
          />
        </td>
      </tr>
      <tr>
        <td>
          <button>
            Save
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
